From: Rik van Riel <riel@redhat.com>
Date: Thu, 27 May 2010 15:40:18 -0400
Subject: [mm] clear page errors when issuing a fresh read of page
Message-id: <20100527114018.76b726ce@annuminas.surriel.com>
Patchwork-id: 25838
O-Subject: [RHEL5 PATCH] do_generic_file_read: clear page errors when issuing
	a fresh read of the page
Bugzilla: 590763
RH-Acked-by: Larry Woodman <lwoodman@redhat.com>
RH-Acked-by: Josef Bacik <josef@redhat.com>
RH-Acked-by: Anton Arapov <Anton@redhat.com>

Fixes bug 590763

Upstream commit 91803b499cca2fe558abad709ce83dc896b80950

From: Jeff Moyer <jmoyer@redhat.com>
Subject: do_generic_file_read: clear page errors when issuing a fresh read of the page

I/O errors can happen due to temporary failures, like multipath
errors or losing network contact with the iSCSI server. Because
of that, the VM will retry readpage on the page.

However, do_generic_file_read does not clear PG_error.  This
causes the system to be unable to actually use the data in the
page cache page, even if the subsequent readpage completes
successfully!

The function filemap_fault has had a ClearPageError before
readpage forever.  This patch simply adds the same to
do_generic_file_read.

Signed-off-by: Rik van Riel <riel@redhat.com>
Signed-off-by: Jeff Moyer <jmoyer@redhat.com>

diff --git a/mm/filemap.c b/mm/filemap.c
index f002f4c..fd7ca9f 100644
--- a/mm/filemap.c
+++ b/mm/filemap.c
@@ -1032,6 +1032,12 @@ page_not_up_to_date:
 		}
 
 readpage:
+		/*
+		 * A previous I/O error may have been due to temporary
+		 * failures, eg. multipath errors.
+		 * PG_error will be set again if readpage fails.
+		 */
+		ClearPageError(page);
 		/* Start the actual read. The read will unlock the page. */
 		error = mapping->a_ops->readpage(filp, page);
 
